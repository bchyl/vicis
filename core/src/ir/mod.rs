pub mod function;
pub mod module;
pub mod types;
pub mod util;
pub mod value;
