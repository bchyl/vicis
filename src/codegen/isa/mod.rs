pub mod x86_64;

use crate::{
    codegen::{
        call_conv::CallConvKind,
        function::instruction::InstructionData,
        lower,
        module::Module,
        register::{RegisterClass, RegisterInfo},
    },
    ir::types::{TypeId, Types},
};
use anyhow::Result;

pub trait TargetIsa: Copy {
    type InstData: InstructionData;
    type RegClass: RegisterClass;
    type RegInfo: RegisterInfo;
    type Lower: lower::Lower<Self>;

    fn module_pass_list() -> Vec<fn(&mut Module<Self>) -> Result<()>>;
    fn default_call_conv() -> CallConvKind;
    fn type_size(types: &Types, ty: TypeId) -> u32;
}
