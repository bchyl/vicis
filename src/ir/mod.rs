pub mod function;
pub mod instruction;
pub mod module;
pub mod types;
pub mod util;
pub mod value;
