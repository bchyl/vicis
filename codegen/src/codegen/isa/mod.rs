pub mod x86_64;

use crate::codegen::{
    call_conv::CallConvKind,
    function::instruction::TargetInst,
    lower,
    module::Module,
    register::{RegisterClass, RegisterInfo},
};
use anyhow::Result;
use vicis_core::ir::{
    module::data_layout::DataLayout,
    types::{Type, Types},
};

pub trait TargetIsa: Clone {
    type Inst: TargetInst;
    type RegClass: RegisterClass;
    type RegInfo: RegisterInfo;
    type Lower: lower::Lower<Self>;

    fn module_passes() -> Vec<for<'a, 'b> fn(&'b mut Module<'a, Self>) -> Result<()>>; // TODO: Implement a pass manager for machine modules
    fn default_call_conv() -> CallConvKind;
    fn type_size(types: &Types, ty: Type) -> u32; // TODO: FIXME: DataLayout can replace this.
    fn data_layout(&self) -> &DataLayout;
}
