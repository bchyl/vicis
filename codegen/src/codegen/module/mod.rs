use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use rustc_hash::FxHashMap;
use std::{fmt, marker::PhantomData};
use vicis_core::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Module as IrModule,
    },
    types::Types,
};

pub struct Module<'a, T: TargetIsa> {
    pub module: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    pub _isa: PhantomData<fn() -> T>,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.module.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.module.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.module.target().triple())?;
        writeln!(f)?;
        for gv in self.global_variables.values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in &self.attributes {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
