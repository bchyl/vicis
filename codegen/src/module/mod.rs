use super::{function::Function, isa::TargetIsa};
use id_arena::Arena;
use std::fmt;
use vicis_core::ir::{module::Module as IrModule, types::Types};

pub struct Module<'a, T: TargetIsa> {
    pub ir: &'a IrModule,
    pub functions: Arena<Function<'a, T>>,
    pub types: Types,
    pub isa: &'a T,
}

impl<T: TargetIsa> fmt::Debug for Module<'_, T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "source_filename = \"{}\"", self.ir.source_filename())?;
        writeln!(
            f,
            "target datalayout = \"{}\"",
            self.ir.target().datalayout()
        )?;
        writeln!(f, "target triple = \"{}\"", self.ir.target().triple())?;
        writeln!(f)?;
        for gv in self.ir.global_variables().values() {
            writeln!(f, "{}", gv.to_string(&self.types))?;
        }
        writeln!(f)?;
        for (_, func) in &self.functions {
            writeln!(f, "{:?}", func)?;
        }
        for (id, attrs) in self.ir.attributes() {
            write!(f, "attributes #{} = {{ ", id)?;
            for attr in attrs {
                write!(f, "{:?} ", attr)?;
            }
            writeln!(f, "}}")?
        }
        Ok(())
    }
}
