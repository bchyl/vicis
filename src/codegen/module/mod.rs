use super::{function::Function, isa::TargetIsa};
use crate::ir::{
    module::{
        attributes::Attribute, global_variable::GlobalVariable, name::Name, Target as ModuleTarget,
    },
    types::Types,
};
use id_arena::Arena;
use rustc_hash::FxHashMap;

pub struct Module<T: TargetIsa> {
    pub name: String,
    pub source_filename: String,
    pub target: ModuleTarget, // TODO
    pub functions: Arena<Function<T>>,
    pub attributes: FxHashMap<u32, Vec<Attribute>>,
    pub global_variables: FxHashMap<Name, GlobalVariable>,
    pub types: Types,
    // TODO: Metadata
    pub arch: T, // TODO: Should be named 'target'
}
