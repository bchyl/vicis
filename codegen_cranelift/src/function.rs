use crate::{instruction, LowerCtx};
use cranelift::{
    codegen::binemit::{NullStackMapSink, NullTrapSink},
    frontend::{FunctionBuilder, FunctionBuilderContext},
    prelude::AbiParam,
};
use cranelift_codegen::Context;
use cranelift_module::{FuncId, Linkage, Module};
use rustc_hash::FxHashMap;
use vicis_core::ir::{
    function::{Function, FunctionId},
    module::Module as LlvmModule,
};

/// Compiles a llvm function to a cranelift function.
pub fn compile_function<M: Module>(
    clif_mod: &mut M,
    clif_ctx: &mut Context,
    llvm_mod: &LlvmModule,
    llvm_func_id: FunctionId,
) {
    let llvm_func = &llvm_mod.functions()[llvm_func_id];
    let mut builder_ctx = FunctionBuilderContext::new();
    let mut lower_ctx = LowerCtx::new(llvm_mod, clif_mod);

    for param in &llvm_func.params {
        let clif_ty = lower_ctx.into_clif_ty(param.ty);
        clif_ctx.func.signature.params.push(AbiParam::new(clif_ty));
    }
    let clif_result_ty = lower_ctx.into_clif_ty(llvm_func.result_ty);
    clif_ctx
        .func
        .signature
        .returns
        .push(AbiParam::new(clif_result_ty));

    compile_body(&mut lower_ctx, &mut builder_ctx, clif_ctx, llvm_func);
}

/// An utility function to declare and define a cranelift function.
pub fn declare_and_define_function<M: Module>(clif_mod: &mut M, clif_ctx: &mut Context) -> FuncId {
    let id = clif_mod
        .declare_function("func", Linkage::Export, &clif_ctx.func.signature)
        .unwrap();
    clif_mod
        .define_function(id, clif_ctx, &mut NullTrapSink {}, &mut NullStackMapSink {})
        .unwrap();
    clif_mod.clear_context(clif_ctx);
    id
}

fn compile_body<M: Module>(
    lower_ctx: &mut LowerCtx<'_, M>,
    builder_ctx: &mut FunctionBuilderContext,
    cl_ctx: &mut Context,
    llvm_func: &Function,
) {
    let mut builder = FunctionBuilder::new(&mut cl_ctx.func, builder_ctx);
    let mut compiler = instruction::InstCompiler {
        lower_ctx,
        llvm_func,
        builder: &mut builder,
        blocks: FxHashMap::default(),
        insts: FxHashMap::default(),
    };

    for (i, block_id) in llvm_func.layout.block_iter().enumerate() {
        let block = compiler.create_block_for(block_id);
        if i == 0 {
            compiler
                .builder
                .append_block_params_for_function_params(block);
        }
        compiler.builder.switch_to_block(block);
        compiler.builder.seal_block(block);
        for inst_id in llvm_func.layout.inst_iter(block_id) {
            compiler.compile(inst_id);
        }
    }

    compiler.builder.finalize();
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn compile_func() {
        use cranelift::prelude::Configurable;
        use cranelift_codegen::isa::CallConv;
        use cranelift_codegen::{isa, settings};
        use cranelift_object::{ObjectBuilder, ObjectModule};
        use vicis_core::ir::module;

        let source = r#"
define dso_local i32 @main() {
  ret i32 42
}"#;

        let mut flag_builder = settings::builder();
        flag_builder.enable("is_pic").unwrap();
        #[cfg(all(target_os = "linux", target_arch = "x86_64"))]
        let isa_builder = isa::lookup_by_name("x86_64-unknown-unknown-elf").unwrap();
        #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
        let isa_builder = isa::lookup_by_name("aarch64-apple-darwin").unwrap();
        let isa = isa_builder.finish(settings::Flags::new(flag_builder));

        let builder = ObjectBuilder::new(
            isa,
            "".to_owned(), // TODO: This will be embedded in the object file.
            cranelift_module::default_libcall_names(),
        )
        .unwrap();
        let mut clif_mod = ObjectModule::new(builder);
        let mut clif_ctx = clif_mod.make_context();

        let module = module::parse_assembly(source).unwrap();
        let llvm_func_id = module.find_function_by_name("main").unwrap();
        compile_function(&mut clif_mod, &mut clif_ctx, &module, llvm_func_id);

        // TODO: FIXME: Depending on OS and ISA, the calling convention may vary.
        // This makes it difficult to do testing using insta because the text representation
        // of a cranelift function contains the name of calling convention.
        // So we have to reset the calling convention here.
        clif_ctx.func.signature.call_conv = CallConv::Fast;
        insta::assert_display_snapshot!(clif_ctx.func.display());
    }

    #[cfg(target_os = "linux")]
    #[test]
    fn compile_ret_42() {
        use cranelift_jit::{JITBuilder, JITModule};
        use std::mem::transmute;
        use vicis_core::ir::module;

        let source = r#"
define dso_local i32 @main() {
  ret i32 42
}"#;

        let builder = JITBuilder::new(cranelift_module::default_libcall_names());
        let mut clif_mod = JITModule::new(builder);
        let mut clif_ctx = clif_mod.make_context();

        let module = module::parse_assembly(source).unwrap();
        let func_id = module.find_function_by_name("main").unwrap();
        compile_function(&mut clif_mod, &mut clif_ctx, &module, func_id);
        let func_id = declare_and_define_function(&mut clif_mod, &mut clif_ctx);
        clif_mod.finalize_definitions();

        let code = clif_mod.get_finalized_function(func_id);
        let code_fn = unsafe { transmute::<_, fn() -> i32>(code) };
        assert_eq!(code_fn(), 42);
    }

    #[cfg(target_os = "linux")]
    #[test]
    fn compile_add() {
        use cranelift_jit::{JITBuilder, JITModule};
        use std::mem::transmute;
        use vicis_core::ir::module;

        let source = r#"
define dso_local i32 @main(i32 %arg.0) {
  %result = add nsw i32 %arg.0, 1
  ret i32 %result
}"#;

        let builder = JITBuilder::new(cranelift_module::default_libcall_names());
        let mut clif_mod = JITModule::new(builder);
        let mut clif_ctx = clif_mod.make_context();

        let module = module::parse_assembly(source).unwrap();
        let func_id = module.find_function_by_name("main").unwrap();
        compile_function(&mut clif_mod, &mut clif_ctx, &module, func_id);
        let id = declare_and_define_function(&mut clif_mod, &mut clif_ctx);
        clif_mod.finalize_definitions();

        let code = clif_mod.get_finalized_function(id);
        let code_fn = unsafe { transmute::<_, fn(i32) -> i32>(code) };
        assert_eq!(code_fn(41), 42);
    }
}
