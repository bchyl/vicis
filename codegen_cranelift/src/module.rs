use crate::{
    function::{compile_function, declare_and_define_function},
    LowerCtx,
};
use cranelift_codegen::Context;
use cranelift_module::Module;
use vicis_core::ir::module::Module as LlvmModule;

pub fn compile_module<M: Module>(clif_mod: &mut M, clif_ctx: &mut Context, llvm_mod: &LlvmModule) {
    let mut lower_ctx = LowerCtx::new(llvm_mod, clif_mod);

    let mut funcs = vec![];
    let mut protos = vec![];

    for func in llvm_mod.functions() {
        if func.1.is_prototype() {
            protos.push(func);
        } else {
            funcs.push(func);
        }
    }

    // Declare prototypes first.
    for (func_id, _) in protos {
        compile_function(&mut lower_ctx, clif_ctx, func_id);
        lower_ctx.clif_mod.clear_context(clif_ctx);
    }

    // Define functions.
    for (func_id, func) in funcs {
        compile_function(&mut lower_ctx, clif_ctx, func_id);
        declare_and_define_function(lower_ctx.clif_mod, clif_ctx, func.name().as_str());
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test1() {
        compile(
            r#"
        declare dso_local i32 @putchar(i8 signext)
        define dso_local i32 @main() {
          call i32 @putchar(i8 signext 65)
          ret i32 0
        }"#,
        );
    }

    fn compile(source: &str) {
        use cranelift::prelude::Configurable;
        use cranelift_codegen::{isa, settings};
        use cranelift_object::{ObjectBuilder, ObjectModule};
        use vicis_core::ir::module::Module;

        let mut flag_builder = settings::builder();
        flag_builder.enable("is_pic").unwrap();
        #[cfg(all(target_os = "linux", target_arch = "aarch64"))]
        let isa_builder = isa::lookup_by_name("aarch64-unknown-unknown-elf").unwrap();
        #[cfg(all(target_os = "linux", target_arch = "x86_64"))]
        let isa_builder = isa::lookup_by_name("x86_64-unknown-unknown-elf").unwrap();
        #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
        let isa_builder = isa::lookup_by_name("aarch64-apple-darwin").unwrap();
        let isa = isa_builder.finish(settings::Flags::new(flag_builder));

        let builder = ObjectBuilder::new(
            isa,
            "".to_owned(), // TODO: This will be embedded in the object file.
            cranelift_module::default_libcall_names(),
        )
        .unwrap();
        let mut clif_mod = ObjectModule::new(builder);
        let mut clif_ctx = clif_mod.make_context();

        let module = Module::try_from(source).unwrap();
        compile_module(&mut clif_mod, &mut clif_ctx, &module);

        clif_mod.finish();
    }
}
