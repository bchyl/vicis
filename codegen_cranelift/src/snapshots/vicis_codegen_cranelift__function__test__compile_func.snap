---
source: codegen_cranelift/src/function.rs
assertion_line: 126
expression: clif_ctx.func.display(None)

---
function u0:0() -> i32 fast {
block0:
    v0 = iconst.i32 42
    return v0
}

